src/createActionListenerMiddleware.ts

```diff
@@ -0,0 +1,358 @@
+import { Middleware, Dispatch, AnyAction, MiddlewareAPI, Action } from 'redux'
+import { TypedActionCreator } from './mapBuilders'
+import { createAction, BaseActionCreator } from './createAction'
+import { WithMiddlewareType } from './tsHelpers'
+
+export type When = 'before' | 'after' | undefined
+type WhenFromOptions<
+  O extends ActionListenerOptions<any, any, any, any>
+> = O extends ActionListenerOptions<any, any, any, infer W> ? W : never
+
+/**
+ * @alpha
+ */
+export interface ActionListenerMiddlewareAPI<
+  S,
+  D extends Dispatch<AnyAction>,
+  O extends ActionListenerOptions<any, any, any, any>
+> extends MiddlewareAPI<D, S> {
+  stopPropagation: WhenFromOptions<O> extends 'after' ? undefined : () => void
+}
+
+/**
+ * @alpha
+ */
+export type ActionListener<
+  A extends AnyAction,
+  S,
+  D extends Dispatch<AnyAction>,
+  O extends ActionListenerOptions<any, any, any, any>
+> = (action: A, api: ActionListenerMiddlewareAPI<S, D, O>) => void
+
+export interface ActionListenerOptions<
+  A extends AnyAction,
+  S,
+  _ extends Dispatch<AnyAction>,
+  W extends When = 'before'
+> {
+  /**
+   * Indicates that the listener should be removed after if has run once.
+   */
+  once?: boolean
+  /**
+   * A function that determines if the listener should run, depending on the action and probably the state.
+   */
+  condition?(action: A, getState: () => S): boolean
```

Comment:
Is there a particular feature this is satisfying that just having the condition around the body of the listener function wouldn't also provide?

```js
function myCondition(action, getState) {
  // ...
}

function myListener(action, { getState } ) {
  // ...
}

store.dispatch(addListener(myListener, { condition: myCondition }))
```
 vs. 

```js
function myCondition(action, getState) {
  // ...
}

function myListener(action, { getState } ) {
  if (myCondition(action, getState)) {
    // ...
  }
}

store.dispatch(addListener(myListener))
```

Similarly to reducers, a higher-order listener could be used it the same condition logic was needed or if the listener was being provided by a third party and I wanted to add a condition to it

```js
import someListener from 'some-dep'

function myListener(action, ...args) {
  if (action.type !== 'secret action type') {
    return someListener(action, ...args)
  }
}

store.dispatch(addListener(myListener))
```

I just feels like it's adding complexity to the internals without provide anything they can do (and relatively easily) in userland... unless I'm missing something?

Reply:
Hmm. Yeah, I could see it being straightforward to do conditional and `once` handling on your own, although we might need to document that as recipes.

Reply:
Hm. I'd really like to keep the `once` option, as having a "react to only the next action of type X" seems really useful sometimes. 
In that context, the condition also has it's value. So I'm not really sure if I would want to let got of that.

Reply:
An alternative option to `once` *could* be to add an `unsubscribe` function to the api. Not 100% if I like the idea yet. Thoughts?

Reply:
Can the conditional `once` functionality not be implemented as

```js
function myListener(action, { getState, dispatch } ) {
  if (myCondition(action, getState)) {
    dispatch(removeListener(myListener))
    // ...
  }
}
```

Reply:
Yes, but that's a lot less obvious to the user than 
```ts
function myListener(action, { getState, dispatch, unsubscribe } ) {
  if (myCondition(action, getState)) {
    unsubscribe()
  }
}
```
so I'd rather go that way, as it doesn't really cost much to add that.